File: human_review/patterns.py
"""Trust patterns taxonomy for human-review.

This module defines the core trust patterns that can be recognized and
auto-approved during code review. These patterns represent mechanical,
patterned changes that don't require human judgment.

Key insight: We're not categorizing ALL changes. We're identifying changes
that fit known trustable patterns. Everything else needs review.
"""

import fnmatch
from dataclasses import dataclass


@dataclass(frozen=True)
class TrustPattern:
    """Definition of a trust pattern."""

    id: str  # Pattern identifier (e.g., "imports:added")
    description: str  # Human-readable description


# Core trust patterns taxonomy (~20 patterns)
# These are ONLY for trustable/mechanical/patterned changes.
TRUST_PATTERNS: dict[str, TrustPattern] = {}


def _register(*patterns: TrustPattern) -> None:
    """Register patterns in the global registry."""
    for p in patterns:
        TRUST_PATTERNS[p.id] = p


# Imports
_register(
    TrustPattern("imports:added", "Import statements added"),
    TrustPattern("imports:removed", "Import statements removed"),
    TrustPattern("imports:reordered", "Imports reordered/reorganized"),
)

# Formatting
_register(
    TrustPattern(
        "formatting:whitespace", "Whitespace changes (spaces, tabs, blank lines)"
    ),
    TrustPattern("formatting:line-length", "Line wrapping/length changes"),
    TrustPattern("formatting:style", "Code style (quotes, trailing commas, etc.)"),
)

# Comments
_register(
    TrustPattern("comments:added", "Comments added"),
    TrustPattern("comments:removed", "Comments removed"),
    TrustPattern("comments:modified", "Comments changed"),
)

# Types & Annotations
_register(
    TrustPattern("types:added", "Type annotations added (no logic change)"),
    TrustPattern("types:removed", "Type annotations removed"),
    TrustPattern("types:modified", "Type annotations changed"),
)

# Files
_register(
    TrustPattern("file:deleted", "File deleted entirely"),
    TrustPattern("file:renamed", "File renamed (content unchanged)"),
    TrustPattern("file:moved", "File moved to different directory"),
)

# Code Movement & Renames (unchanged logic)
_register(
    TrustPattern(
        "code:relocated",
        "Code relocated with no behavior change (reordering, not new class/scope)",
    ),
    TrustPattern("rename:variable", "Variable/constant renamed"),
    TrustPattern("rename:function", "Function renamed"),
    TrustPattern("rename:class", "Class renamed"),
    TrustPattern("rename:parameter", "Parameter renamed"),
)

# Generated & Mechanical
_register(
    TrustPattern(
        "generated:lockfile", "Package lock file (package-lock.json, uv.lock, etc.)"
    ),
    TrustPattern("generated:config", "Auto-generated configuration"),
    TrustPattern("generated:migration", "Database migration files"),
    TrustPattern("version:bumped", "Version number changed"),
)

# Removal
_register(
    TrustPattern("remove:deprecated", "Deprecated code removed"),
)


def get_pattern(pattern_id: str) -> TrustPattern | None:
    """Get a pattern by ID."""
    return TRUST_PATTERNS.get(pattern_id)


def is_valid_pattern(pattern_id: str) -> bool:
    """Check if a pattern ID is valid (exists in taxonomy or is custom:*)."""
    if pattern_id in TRUST_PATTERNS:
        return True
    # Allow custom patterns (custom:whatever)
    if pattern_id.startswith("custom:"):
        return True
    return False


def get_all_patterns() -> list[TrustPattern]:
    """Get all registered patterns."""
    return list(TRUST_PATTERNS.values())


def get_category(pattern_id: str) -> str:
    """Extract the category from a pattern ID (the part before the colon)."""
    if ":" in pattern_id:
        return pattern_id.split(":")[0]
    return pattern_id


def pattern_matches_glob(pattern_id: str, glob_pattern: str) -> bool:
    """Check if a pattern ID matches a glob pattern.

    Examples:
        pattern_matches_glob("imports:added", "imports:*") -> True
        pattern_matches_glob("imports:added", "imports:added") -> True
        pattern_matches_glob("imports:added", "*:added") -> True
        pattern_matches_glob("imports:added", "formatting:*") -> False
    """
    return fnmatch.fnmatch(pattern_id, glob_pattern)


def patterns_match_trust_list(
    patterns: list[str], trust_list: list[str]
) -> tuple[bool, list[str]]:
    """Check if all patterns are trusted by the trust list.

    Args:
        patterns: List of pattern IDs from a hunk (e.g., ["imports:added"])
        trust_list: User's configured trust patterns (e.g., ["imports:*", "formatting:*"])

    Returns:
        (all_trusted, untrusted_patterns)
        - all_trusted: True if every pattern in `patterns` matches at least one in trust_list
        - untrusted_patterns: List of patterns that weren't matched
    """
    if not patterns:
        # Empty patterns = needs review (no trustable pattern recognized)
        return False, []

    untrusted = []
    for pattern in patterns:
        is_trusted = any(
            pattern_matches_glob(pattern, trusted) for trusted in trust_list
        )
        if not is_trusted:
            untrusted.append(pattern)

    return len(untrusted) == 0, untrusted


def format_pattern_list(patterns: list[str]) -> str:
    """Format a list of patterns for display."""
    if not patterns:
        return "(no patterns)"
    return ", ".join(patterns)


def get_pattern_description(pattern_id: str) -> str:
    """Get the description for a pattern ID."""
    pattern = TRUST_PATTERNS.get(pattern_id)
    if pattern:
        return pattern.description
    if pattern_id.startswith("custom:"):
        return f"Custom pattern: {pattern_id[7:]}"
    return pattern_id

File: vscode/src/trust/patterns.ts
/**
 * Trust patterns taxonomy for human-review.
 * Ported from Python human_review/patterns.py
 *
 * This module defines the core trust patterns that can be recognized and
 * auto-approved during code review. These patterns represent mechanical,
 * patterned changes that don't require human judgment.
 *
 * Key insight: We're not categorizing ALL changes. We're identifying changes
 * that fit known trustable patterns. Everything else needs review.
 */

/**
 * Definition of a trust pattern.
 */
export interface TrustPattern {
  id: string;
  description: string;
}

/**
 * Core trust patterns taxonomy (~20 patterns).
 * These are ONLY for trustable/mechanical/patterned changes.
 */
export const TRUST_PATTERNS: Record<string, TrustPattern> = {
  // Imports
  "imports:added": { id: "imports:added", description: "Import statements added" },
  "imports:removed": { id: "imports:removed", description: "Import statements removed" },
  "imports:reordered": { id: "imports:reordered", description: "Imports reordered/reorganized" },

  // Formatting
  "formatting:whitespace": {
    id: "formatting:whitespace",
    description: "Whitespace changes (spaces, tabs, blank lines)",
  },
  "formatting:line-length": {
    id: "formatting:line-length",
    description: "Line wrapping/length changes",
  },
  "formatting:style": {
    id: "formatting:style",
    description: "Code style (quotes, trailing commas, etc.)",
  },

  // Comments
  "comments:added": { id: "comments:added", description: "Comments added" },
  "comments:removed": { id: "comments:removed", description: "Comments removed" },
  "comments:modified": { id: "comments:modified", description: "Comments changed" },

  // Types & Annotations
  "types:added": { id: "types:added", description: "Type annotations added (no logic change)" },
  "types:removed": { id: "types:removed", description: "Type annotations removed" },
  "types:modified": { id: "types:modified", description: "Type annotations changed" },

  // Files
  "file:deleted": { id: "file:deleted", description: "File deleted entirely" },
  "file:renamed": { id: "file:renamed", description: "File renamed (content unchanged)" },
  "file:moved": { id: "file:moved", description: "File moved to different directory" },

  // Code Movement & Renames (unchanged logic)
  "code:relocated": {
    id: "code:relocated",
    description: "Code relocated with no behavior change (reordering, not new class/scope)",
  },
  "rename:variable": { id: "rename:variable", description: "Variable/constant renamed" },
  "rename:function": { id: "rename:function", description: "Function renamed" },
  "rename:class": { id: "rename:class", description: "Class renamed" },
  "rename:parameter": { id: "rename:parameter", description: "Parameter renamed" },

  // Generated & Mechanical
  "generated:lockfile": {
    id: "generated:lockfile",
    description: "Package lock file (package-lock.json, uv.lock, etc.)",
  },
  "generated:config": { id: "generated:config", description: "Auto-generated configuration" },
  "generated:migration": { id: "generated:migration", description: "Database migration files" },
  "version:bumped": { id: "version:bumped", description: "Version number changed" },

  // Removal
  "remove:deprecated": { id: "remove:deprecated", description: "Deprecated code removed" },
};

/**
 * Get a pattern by ID.
 */
export function getPattern(patternId: string): TrustPattern | undefined {
  return TRUST_PATTERNS[patternId];
}

/**
 * Check if a pattern ID is valid (exists in taxonomy or is custom:*).
 */
export function isValidPattern(patternId: string): boolean {
  if (patternId in TRUST_PATTERNS) {
    return true;
  }
  // Allow custom patterns (custom:whatever)
  if (patternId.startsWith("custom:")) {
    return true;
  }
  return false;
}

/**
 * Get all registered patterns.
 */
export function getAllPatterns(): TrustPattern[] {
  return Object.values(TRUST_PATTERNS);
}

/**
 * Extract the category from a pattern ID (the part before the colon).
 */
export function getCategory(patternId: string): string {
  if (patternId.includes(":")) {
    return patternId.split(":")[0];
  }
  return patternId;
}

/**
 * Get the description for a pattern ID.
 */
export function getPatternDescription(patternId: string): string {
  const pattern = TRUST_PATTERNS[patternId];
  if (pattern) {
    return pattern.description;
  }
  if (patternId.startsWith("custom:")) {
    return `Custom pattern: ${patternId.slice(7)}`;
  }
  return patternId;
}

/**
 * Format a list of patterns for display.
 */
export function formatPatternList(patterns: string[]): string {
  if (patterns.length === 0) {
    return "(no patterns)";
  }
  return patterns.join(", ");
}
